File: archive/speedgrader/app/src/main/java/com/instructure/speedgrader/activities/LoginActivity.java
/*
 * Copyright (C) 2016 - present  Instructure, Inc.
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU General Public License as published by
 *     the Free Software Foundation, version 3 of the License.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU General Public License for more details.
 *
 *     You should have received a copy of the GNU General Public License
 *     along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

package com.instructure.speedgrader.activities;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.graphics.drawable.Drawable;
import android.net.Uri;
import android.os.Bundle;
import android.widget.ImageView;
import android.widget.Toast;

import com.crashlytics.android.Crashlytics;
import com.instructure.canvasapi.utilities.APIHelpers;
import com.instructure.pandautils.utils.CanvasContextColor;
import com.instructure.speedgrader.R;
import com.instructure.canvasapi.model.User;
import com.instructure.canvasapi.utilities.ErrorDelegate;
import com.instructure.loginapi.login.URLSignIn;
import com.instructure.speedgrader.util.App;
import com.instructure.speedgrader.util.CanvasErrorDelegate;
import io.fabric.sdk.android.Fabric;


public class LoginActivity extends URLSignIn {

    Intent routingIntent;
    @Override
    public void onCreate(Bundle savedInstanceState) {
        Intent intent = getIntent();
        if (intent != null) {
            Bundle bundle = intent.getExtras();
            if (bundle != null && bundle.containsKey(com.instructure.loginapi.login.util.Const.ROUTING_CANVAS_CONEXT)
                    && bundle.containsKey(com.instructure.loginapi.login.util.Const.ROUTING_ASSINGMENT)) {
                routingIntent = intent;
            }
        }

        super.onCreate(savedInstanceState);
        final Drawable connectArrow = CanvasContextColor.getColoredDrawable(this, R.drawable.ic_cv_login_arrow, getResources().getColor(R.color.sg_defaultPrimary));
        ((ImageView)findViewById(com.instructure.loginapi.login.R.id.connect)).setImageDrawable(connectArrow);
    }

    public ErrorDelegate getErrorDelegate() {
        return new CanvasErrorDelegate();
    }

    public void startNextActivity() {
        if(((App)getApplication()).shouldShowTutorial()){
            startActivity(new Intent(LoginActivity.this, TutorialActivity.class));
            return;
        }
        if(routingIntent != null){
            startActivity(HomeActivity.createIntent(getApplicationContext(), routingIntent.getExtras()));
        } else{
            startActivity(HomeActivity.createIntent(getApplicationContext()));
        }

    }

    public void startNextActivity(Uri passedURI) {
        if(routingIntent != null){
            startActivity(HomeActivity.createIntent(getApplicationContext(), routingIntent.getExtras()));
        } else{
            startActivity(HomeActivity.createIntent(getApplicationContext(), passedURI));
        }
    }

    @Override
    public int getRootLayout() {
        return R.layout.sg_url_sign_in;
    }

    @Override
    public boolean shouldShowHelpButton() {
        return false;
    }

    public void initializeLoggingForUser(boolean isSimonFraser, User signedInUser) {
        if (!isSimonFraser) {
            Crashlytics.setUserIdentifier(Long.toString(signedInUser.getId()));
        } else {
            //Clear context for Crashlytics.
            Crashlytics.setUserIdentifier("");
        }
        Crashlytics.setString("domain", APIHelpers.getDomain(LoginActivity.this));

    }

    public void startCrashlytics() {
        Fabric.with(this, new Crashlytics());
    }

    public void startHelpShift() {}

    public void startGoogleAnalytics() {}

    public void showHelpShiftSupport() {}

    public void trackAppFlow(Activity activity) {}

    public void displayMessage(String message, int messageType) {
        Toast.makeText(LoginActivity.this, message, Toast.LENGTH_SHORT).show();
    }

    public void handleNightlyBuilds() {}

    public void refreshWidgets() {}

    public void deleteCachedFiles() {}

    @Override
    public String getPrefsFileName() {
        return App.PREF_FILE_NAME;
    }

    @Override
    public String getPrefsPreviousDomainKey() {
        return App.PREF_NAME_PREVIOUS_DOMAINS;
    }

    @Override
    public String getPrefsOtherSignedInUsersKey() {
        return App.OTHER_SIGNED_IN_USERS_PREF_NAME;
    }

    @Override
    public String getPrefsMultiUserKey() {
        return App.MULTI_SIGN_IN_PREF_NAME;
    }

    //Intent
    public static Intent createIntent(Context context) {
        Intent intent = new Intent(context, LoginActivity.class);
        return intent;
    }

    public static Intent createIntent(Context context, boolean showMessage, String message) {
        Intent intent = new Intent(context, LoginActivity.class);
        intent.putExtra("showMessage", showMessage);
        intent.putExtra("messageToUser", message);
        return intent;
    }
}

File: apps/parent/src/main/java/com/instructure/parentapp/features/login/LoginActivity.kt
/*
 * Copyright (C) 2024 - present Instructure, Inc.
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU General Public License as published by
 *     the Free Software Foundation, version 3 of the License.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU General Public License for more details.
 *
 *     You should have received a copy of the GNU General Public License
 *     along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

package com.instructure.parentapp.features.login

import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import androidx.core.content.ContextCompat
import com.instructure.canvasapi2.models.User
import com.instructure.canvasapi2.utils.ApiPrefs
import com.instructure.loginapi.login.activities.BaseLoginInitActivity
import com.instructure.loginapi.login.util.QRLogin
import com.instructure.pandautils.utils.AppType
import com.instructure.pandautils.utils.Const
import com.instructure.pandautils.utils.Utils
import com.instructure.parentapp.R
import com.instructure.parentapp.features.login.routevalidator.RouteValidatorActivity
import com.instructure.parentapp.features.main.MainActivity
import dagger.hilt.android.AndroidEntryPoint

@AndroidEntryPoint
class LoginActivity : BaseLoginInitActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        // Check to see if we are switching users from an external QR scan
        if (QRLogin.verifySSOLoginUri(intent.data, AppType.PARENT)) {
            startActivity(RouteValidatorActivity.createIntent(this, intent.data!!))
            finish()
        }
    }

    override fun beginLoginFlowIntent(): Intent {
        return LoginLandingPageActivity.createIntent(this)
    }

    override fun themeColor(): Int {
        return ContextCompat.getColor(this, R.color.login_parentAppTheme)
    }

    override fun userAgent() = Const.PARENT_USER_AGENT

    /**
     * ONLY USE FOR UI TESTING
     * Skips the traditional login process by directly setting the domain, token, and user info.
     */
    fun loginWithToken(token: String, domain: String, user: User) {
        ApiPrefs.accessToken = token
        ApiPrefs.domain = domain
        ApiPrefs.user = user
        ApiPrefs.userAgent = Utils.generateUserAgent(this, userAgent())
        val intent = MainActivity.createIntent(this, Uri.EMPTY)
        finish()
        startActivity(intent)
    }

    companion object {
        fun createIntent(context: Context): Intent {
            val intent = Intent(context, LoginActivity::class.java)
            intent.flags = Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TASK
            return intent
        }

        // Used specifically for a QR Scan user switch
        fun createIntent(context: Context, uri: Uri): Intent {
            val intent = Intent(context, LoginActivity::class.java)
            intent.flags = Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TASK
            intent.data = uri
            return intent
        }
    }
}
